buck/config.py

```diff
@@ -0,0 +1,654 @@
+        except AttributeError:
+            return False
+
+    def __str__(self) -> str:
+        return self._name
+
+    def __repr__(self) -> str:
+        return "default"
+
+
+default = Default()
+
+
+def get_envs_singleton() -> Dict[str, Env]:
+    """Return the ENVS singleton."""
+    global envs
+    if envs is None:
+        envs = {}
+    return envs
+
+
+def get_selectors_singleton() -> Selectors:
+    global selectors
+    if selectors is None:
+        selectors = {}
+    return selectors
+
+
+def get_mappings_singleton() -> Mappings:
+    global mappings
+    if mappings is None:
+        mappings = {}
+    return mappings
+
+
+def validate_env_vars(**data: EnvValuesType) -> Dict[str, str]:
+    """Validate that the keys provided in data are of the correct type.
+
+    If the env_name prefix is 'testenv' then the keys are verified; otherwise
+    the type is detected and then converted to a string.  Any iterables are
+    assummed to be joined by a newline.
+
+    Note that if the data keys include 'pass_env' or 'set_env' then they are
+    rationalised to 'passenv' and 'setenv'; they must not be duplicated.
+
+    :param data: the key, values that make up the an env.
+    :raises: TypeError if the value doesn't conform to the valid type.
+    :raises: KeyError on an unknown key.
+    """
+    type_checks = dict(
+        env_name=is_str,
+        description=is_str,
+        setenv=partial(is_str_or_iterable_str, ', '),
+        commands=partial(is_str_or_iterable_str, "\n"),
+        allowlist_externals=partial(is_str_or_iterable_str, ', '),
+        passenv=partial(is_str_or_iterable_str, ', '),
+        deps=partial(is_str_or_iterable_str, "\n"),
+        basepython=is_str,
+        platform=is_str,
+        parallel_show_output=is_bool,
+        recreate=is_bool,
+        skip_install=is_bool,
+        labels=partial(is_str_or_iterable_str, ', '),
+    )
+    mapped_data = {}
+
+    # pass_env is an alias for passenv, and set_env is an alias for, so make
+    # sure both of each are not valid.
+    env_keys = data.keys()
+    if 'pass_env' in env_keys and 'passenv' in env_keys:
+        raise ParameterError("Can't have both 'pass_env' and 'passenv'")
+    if 'set_env' in env_keys and 'setenv' in env_keys:
+        raise ParameterError("Can't have both 'set_env' and 'setenv'")
+    if 'pass_env' in env_keys:
+        data['passenv'] = data['pass_env']
+        del data['pass_env']
+    if 'set_env' in env_keys:
+        data['setenv'] = data['set_env']
+        del data['set_env']
+
+    for k, v in data.items():
+        try:
+            mapped_v = type_checks[k](v)
+        except KeyError:
+            raise KeyError(f"Unknown env key '{k}'")
+        mapped_data[k] = mapped_v
+    return mapped_data
+
+
+def is_str_or_iterable_str(separator: str,
+                           # v: Union[str, Iterable[str]]) -> str:
+                           v: EnvValuesType) -> str:
```

Comment:
Well `EnvValuesType` is the type that is passed in, but the function body makes sure it's just a `str` or `Iterable[str]` and otherwise raises a `TypeError` if it's a bool.  So the type is correct.  I need to pull out the comment, though, so I'll do that.